File: vehicle_speed_detection.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/dnn.hpp>
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>

using namespace cv;
using namespace dnn;
using namespace std;

// Function to get output layer names
vector<String> getOutputLayersNames(const Net& net) {
    vector<String> layerNames = net.getLayerNames();
    vector<String> outputLayers;
    for (auto i : net.getUnconnectedOutLayers())
        outputLayers.push_back(layerNames[i - 1]);
    return outputLayers;
}

int main() {
    // Paths to YOLO files
    string weightsPath = "./yolo/yolov4.weights";   // Update with your YOLO weights path
    string configPath = "./yolo/yolov4.cfg";        // Update with your YOLO config path
    string namesPath = "./yolo/coco.names";         // Update with your COCO class labels file path
    string videoPath = "D:/Vehicle speed detection/12691893_3840_2160_30fps.mp4"; // Update with your video path

    // Load YOLO model
    Net net = readNet(weightsPath, configPath);
    net.setPreferableBackend(DNN_BACKEND_OPENCV);
    net.setPreferableTarget(DNN_TARGET_CPU);

    // Load COCO class labels
    vector<string> classes;
    ifstream ifs(namesPath.c_str());
    string line;
    while (getline(ifs, line)) {
        classes.push_back(line);
    }

    // Load video
    VideoCapture cap(videoPath);
    if (!cap.isOpened()) {
        cerr << "Error: Could not open video." << endl;
        return -1;
    }

    // Get video properties
    double fps = cap.get(CAP_PROP_FPS);
    cout << "Video FPS: " << fps << endl;

    // Define a known distance (in meters) for speed calculation
    const double knownDistance = 150.0; // Adjust this based on your real-world reference

    // Initialize tracking variables
    vector<Rect> previousFrameBoxes;

    Mat frame;
    while (cap.read(frame)) {
        // Resize frame for better visibility
        const int resizeWidth = 1280, resizeHeight = 720;
        resize(frame, frame, Size(resizeWidth, resizeHeight));
        int width = frame.cols, height = frame.rows;

        // Create blob from frame
        Mat blob;
        blobFromImage(frame, blob, 0.00392, Size(416, 416), Scalar(), true, false);
        net.setInput(blob);

        // Forward pass
        vector<Mat> outs;
        net.forward(outs, getOutputLayersNames(net));

        // Process detections
        vector<int> classIds;
        vector<float> confidences;
        vector<Rect> boxes;
        for (size_t i = 0; i < outs.size(); ++i) {
            float* data = (float*)outs[i].data;
            for (int j = 0; j < outs[i].rows; ++j, data += outs[i].cols) {
                Mat scores = outs[i].row(j).colRange(5, outs[i].cols);
                Point classIdPoint;
                double confidence;
                minMaxLoc(scores, 0, &confidence, 0, &classIdPoint);
                int classId = classIdPoint.x;

                if (confidence > 0.5 &&
                    (classes[classId] == "car" || classes[classId] == "truck" ||
                     classes[classId] == "bus" || classes[classId] == "motorbike")) {
                    int centerX = int(data[0] * width);
                    int centerY = int(data[1] * height);
                    int w = int(data[2] * width);
                    int h = int(data[3] * height);
                    int x = centerX - w / 2;
                    int y = centerY - h / 2;

                    boxes.push_back(Rect(x, y, w, h));
                    confidences.push_back((float)confidence);
                    classIds.push_back(classId);
                }
            }
        }

        // Non-Maximum Suppression
        vector<int> indices;
        NMSBoxes(boxes, confidences, 0.5, 0.4, indices);

        // Process the detections
        vector<Rect> currentFrameBoxes;
        for (int idx : indices) {
            Rect box = boxes[idx];
            Point vehicleCenter(box.x + box.width / 2, box.y + box.height / 2);

            // Match with previous frame vehicles
            bool matched = false;
            for (Rect prevBox : previousFrameBoxes) {
                Point prevCenter(prevBox.x + prevBox.width / 2, prevBox.y + prevBox.height / 2);
                double distancePixels = norm(vehicleCenter - prevCenter);

                if (distancePixels < 50) { // Adjust threshold
                    matched = true;
                    double timeSeconds = 1.0 / fps;
                    double speedMps = (distancePixels / width) * knownDistance / timeSeconds;
                    double speedKmph = speedMps * 3.6;

                    putText(frame, format("Speed: %.2f km/h", speedKmph),
                            Point(box.x, box.y - 10), FONT_HERSHEY_SIMPLEX, 0.5,
                            Scalar(0, 255, 255), 2);
                    cout << "Vehicle Speed: " << speedKmph << " km/h" << endl;
                }
            }

            currentFrameBoxes.push_back(box);

            // Draw bounding box
            rectangle(frame, box, Scalar(0, 255, 0), 2);
            putText(frame, format("%s: %.2f", classes[classIds[idx]].c_str(), confidences[idx]),
                    Point(box.x, box.y - 10), FONT_HERSHEY_SIMPLEX, 0.5, Scalar(0, 255, 0), 2);
        }

        previousFrameBoxes = currentFrameBoxes;

        // Display the frame
        imshow("Vehicle Speed Detection", frame);

        if (waitKey(1) == 'q') break; // Exit on 'q'
    }

    cap.release();
    destroyAllWindows();

    return 0;
}

File: vehicle_speed_detection.py
import cv2
import numpy as np

# Paths to YOLO files
weights_path = r"./yolo/yolov4.weights"
config_path = r"./yolo/yolov4.cfg"
names_path = r"./yolo/coco.names"

# Load the YOLO model
net = cv2.dnn.readNet(weights_path, config_path)
net.setPreferableBackend(cv2.dnn.DNN_BACKEND_OPENCV)
net.setPreferableTarget(cv2.dnn.DNN_TARGET_CPU)

# Load COCO class labels
with open(names_path, "r") as f:
    classes = [line.strip() for line in f.readlines()]

# Get the output layer names
layer_names = net.getLayerNames()
output_layers = [layer_names[i - 1] for i in net.getUnconnectedOutLayers()]

# Load video
video_path = r"D:\Vehicle speed detection\12691893_3840_2160_30fps.mp4"
cap = cv2.VideoCapture(video_path)

if not cap.isOpened():
    print("Error: Could not open video.")
    exit()

# Video display size
resize_width, resize_height = 1280, 720

# Video FPS
fps = cap.get(cv2.CAP_PROP_FPS)

# Known real-world distance for speed calculation
known_distance = 150  # Adjust this based on your scenario

# Variables for tracking vehicles
previous_frame_boxes = []

while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break

    # Resize frame
    frame = cv2.resize(frame, (resize_width, resize_height))
    height, width = frame.shape[:2]

    # Create a blob from the frame
    blob = cv2.dnn.blobFromImage(frame, 0.00392, (416, 416), (0, 0, 0), True, crop=False)
    net.setInput(blob)
    outs = net.forward(output_layers)

    # Initialize lists for detected bounding boxes, confidences, and centers
    boxes, confidences, class_ids, centers = [], [], [], []

    for out in outs:
        for detection in out:
            scores = detection[5:]
            class_id = np.argmax(scores)
            confidence = scores[class_id]

            if confidence > 0.5 and classes[class_id] in ["car", "truck", "bus", "motorbike"]:
                center_x = int(detection[0] * width)
                center_y = int(detection[1] * height)
                w = int(detection[2] * width)
                h = int(detection[3] * height)

                x = int(center_x - w / 2)
                y = int(center_y - h / 2)

                # Append bounding box coordinates and center
                boxes.append([x, y, w, h])
                confidences.append(float(confidence))
                class_ids.append(class_id)
                centers.append((center_x, center_y))

    # Perform Non-Maximum Suppression (NMS)
    indexes = cv2.dnn.NMSBoxes(boxes, confidences, 0.5, 0.4)

    current_frame_boxes = []
    if len(indexes) > 0:
        for i in indexes.flatten():
            x, y, w, h = boxes[i]
            center_x, center_y = centers[i]

            matched = False
            for prev_box in previous_frame_boxes:
                prev_x, prev_y, prev_w, prev_h, prev_center_x, prev_center_y = prev_box
                distance_pixels = np.linalg.norm(np.array([center_x, center_y]) - np.array([prev_center_x, prev_center_y]))

                if distance_pixels < 50:  # Threshold for matching
                    matched = True
                    time_seconds = 1 / fps
                    speed_mps = (distance_pixels / width) * known_distance / time_seconds
                    speed_kmph = speed_mps * 3.6

                    # Determine direction
                    dx = center_x - prev_center_x
                    dy = center_y - prev_center_y
                    direction = ""
                    if abs(dx) > abs(dy):
                        direction = "Right" if dx > 0 else "Left"
                    else:
                        direction = "Down" if dy > 0 else "Up"

                    # Display speed and direction
                    cv2.putText(frame, f"Speed: {speed_kmph:.2f} km/h, {direction}", (x, y - 30),
                                cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 255), 2)

            # Add the current bounding box with center for tracking
            current_frame_boxes.append([x, y, w, h, center_x, center_y])
            label = f"{classes[class_ids[i]]}: {confidences[i]:.2f}"
            cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 2)
            cv2.putText(frame, label, (x, y - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)

    # Update the previous frame boxes
    previous_frame_boxes = current_frame_boxes

    # Display the frame
    cv2.imshow("Vehicle Speed and Direction Detection", frame)

    # Exit on 'q' key press
    if cv2.waitKey(1) & 0xFF == ord("q"):
        break

cap.release()
cv2.destroyAllWindows()
